C/STrace/EtwLogger.h

```diff
@@ -0,0 +1,314 @@
+	const auto providerMetadataLength = (uint16_t)((strlen(providerName) + 1) + sizeof(uint16_t));
+	const auto providerMetadata = (struct detail::ProviderMetadata*)ExAllocatePoolWithTag(NonPagedPoolNx, providerMetadataLength, 'wteO');
+	if (providerMetadata == NULL)
+	{
+		return EVENT_DATA_DESCRIPTOR{};
+	}
+
+	RtlSecureZeroMemory(providerMetadata, providerMetadataLength);
+	providerMetadata->TotalLength = providerMetadataLength;
+	strcpy(providerMetadata->ProviderName, providerName);
+
+	// Create an EVENT_DATA_DESCRIPTOR pointing to the metadata.
+	EVENT_DATA_DESCRIPTOR providerMetadataDesc;
+	EventDataDescCreate(&providerMetadataDesc, providerMetadata, providerMetadata->TotalLength);
+	providerMetadataDesc.Type = EVENT_DATA_DESCRIPTOR_TYPE_PROVIDER_METADATA;  // Descriptor contains provider metadata.
+
+	return providerMetadataDesc;
+}
+
+size_t SizeOfFieldsMeta()
+{
+	return 0;
+}
+
+template<typename FieldValue, typename... Rest>
+size_t SizeOfFieldsMeta(const char* fieldName, int fieldType, FieldValue& fieldValue, Rest... rest)
+{
+	UNREFERENCED_PARAMETER(fieldType);
+	UNREFERENCED_PARAMETER(fieldValue);
+	return strlen(fieldName) + 1 + sizeof(uint8_t) + SizeOfFieldsMeta(rest...);
+}
+
+void SetFieldMetadata(uint8_t* current)
+{
+	UNREFERENCED_PARAMETER(current);
+}
+
+template<typename FieldValue, typename... Rest>
+void SetFieldMetadata(uint8_t* current, const char* fieldName, int fieldType, FieldValue& fieldValue, Rest... rest)
+{
+	UNREFERENCED_PARAMETER(fieldValue);
+
+	strcpy((char*)current, fieldName);
+	current[strlen(fieldName) + 1] = (uint8_t)fieldType;
+
+	current += strlen(fieldName) + 1 + sizeof(uint8_t);
+	SetFieldMetadata(current, rest...);
+}
+
+template<typename... Arguments>
+EVENT_DATA_DESCRIPTOR CreateEventMetadata(const char* eventName, Arguments... args)
+{
+	// Allocate the total size of the event metadata structure.
+	//
+	// This consists of, in order:
+	//
+	// * The total length of the structure
+	// * The event tag byte (currently always 0)
+	// * The name of the event
+	// * An array of field metadata structures, which are:
+	//     * the name of the field
+	//     * a single byte for the type.
+	const auto eventMetadataHeaderLength = strlen(eventName) + 1 + sizeof(uint16_t) + sizeof(uint8_t);
+	const auto eventMetadataLength = (uint16_t)(eventMetadataHeaderLength + SizeOfFieldsMeta(args...));
+	const auto eventMetadata = (struct detail::EventMetadata*)ExAllocatePoolWithTag(NonPagedPoolNx, eventMetadataLength, 'wteE');
+	if (eventMetadata == NULL)
+	{
+		return EVENT_DATA_DESCRIPTOR{};
+	}
+
+	RtlSecureZeroMemory(eventMetadata, eventMetadataLength);
+
+	// Set the first three fields, metadata about the event.
+	eventMetadata->TotalLength = eventMetadataLength;
+	eventMetadata->Tag = 0;
+	strcpy(eventMetadata->EventName, eventName);
+
+	// Set the metadata for each field.
+	uint8_t* currentLocation = ((uint8_t*)eventMetadata) + eventMetadataHeaderLength;
```

Comment:
make this a char* when doing manual pointer offset math, there are complex aliasing rules in C++ and I only know for sure this is not undefined behavior when using char*